### PR TITLE
refactor!(global-shortcut): better APIs DX

### LINKED ISSUE #1012: the "tauri-plugin-global-shortcut" plugin never removes items when unregistering shortcuts.
Hello, I've gone through the code of the "tauri-plugin-global-shortcut" project. 
I noticed that in the `GlobalShortcut` struct, there's no code found within the `unregister` and `unregister_all` methods to remove items stored in the `shortcuts` property. 
Consequently, this could lead to erroneous returns when querying with the `is_registered` method after calling the `unregister` method.

```
    pub fn register_all<T: TryInto<ShortcutWrapper>, S: IntoIterator<Item = T>>(
        &self,
        shortcuts: S,
    ) -> Result<()>
    where
        T::Error: std::error::Error,
    {
        let mut s = Vec::new();
        for shortcut in shortcuts {
            s.push((try_into_shortcut(shortcut)?, None));
        }
        self.register_all_internal(s, ShortcutSource::Rust)
    }
    pub fn unregister<S: TryInto<ShortcutWrapper>>(&self, shortcut: S) -> Result<()>
    where
        S::Error: std::error::Error,
    {
        acquire_manager(&self.manager)?
            .unregister(try_into_shortcut(shortcut)?)
            .map_err(Into::into)
        //Note:
        //The "self.shortcuts" property is not updated here.
    }

    pub fn unregister_all<T: TryInto<ShortcutWrapper>, S: IntoIterator<Item = T>>(
        &self,
        shortcuts: S,
    ) -> Result<()>
    where
        T::Error: std::error::Error,
    {
        let mut s = Vec::new();
        for shortcut in shortcuts {
            s.push(try_into_shortcut(shortcut)?);
        }
        acquire_manager(&self.manager)?
            .unregister_all(&s)
            .map_err(Into::into)
        //Note:
        //The "self.shortcuts" property is not updated here.
    }
```
[https://github.com/tauri-apps/tauri-plugin-global-shortcut/blob/v2/src/lib.rs](url)

### PR DESCRIPTION
**Pull request by amrbashir**
ref: https://github.com/tauri-apps/plugins-workspace/issues/965

### CODE REVIEW

File: .changes/global-shortcut-refactor.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+---
+"global-shortcut": "patch"
+---
+
+**Breaking change** Refactored the plugin Rust APIs for better DX and flexibility:
+
+- `GlobalShortcutExt::register` and `GlobalShortcutExt::register_all` now takes a second argument, which is a handler to be called when the shortcut is triggered.
+- Add `GlobalShortcutExt::register_wihtout_handler` and `GlobalShortcutExt::register_all_wihtout_handler` that work the same as old APIs (i.e. register without a handler).
```
Comment (simonhyll) [inline .changes/global-shortcut-refactor.md]:
```suggestion
- Add `GlobalShortcutExt::register_without_handler` and `GlobalShortcutExt::register_all_without_handler` that work the same as old APIs (i.e. register without a handler).
```

**Review by simonhyll — CHANGES_REQUESTED** [review]

**Comment by lucasfernog** [comment]
Do we reall need the without_handler variants? can't users just define an empty closure and then the global listener handles the event?

**Comment by amrbashir** [comment]
I am not happy with it either, the idea is to provide a method to register a shortcut without handler that is appealing to use, without using an empty closure, but a long method name is not good either, maybe just call it `register2`?

**Comment by lucasfernog** [comment]
maybe `register_without_handler` keeps its name `register` and the new API is named `on_shortcut`? so it's clear you intend on providing a callback there

that's even a non breaking change I believe

**Comment by amrbashir** [comment]
> maybe `register_without_handler` keeps its name `register` and the new API is named `on_shortcut`? so it's clear you intend on providing a callback there

I am afraid that might confuse users because `on_` methods throughout tauri, are used to register just the listener.

**Comment by amrbashir** [comment]
Changed to take an `Option` instead

**Comment by lucasfernog** [comment]
> > maybe `register_without_handler` keeps its name `register` and the new API is named `on_shortcut`? so it's clear you intend on providing a callback there
> 
> I am afraid that might confuse users because `on_` methods throughout tauri, are used to register just the listener.

that's what the on_ here will do:

```rust
fn on_shortcut(&self, shortcut: &str, handler: impl Fn()) {}
fn register(&self, shortcut: &str) {}
```

**Comment by amrbashir** [comment]
maybe I am overthinking it, let's roll with that and see user feedback

**Comment by amrbashir** [comment]
> that's what the on_ here will do:
> 
> ```rust
> fn on_shortcut(&self, shortcut: &str, handler: impl Fn()) {}
> fn register(&self, shortcut: &str) {}
> ```

hmm if we go with that approach, there will be inconsistencies between the JS and Rust APIs

**Comment by lucasfernog** [comment]
it's already inconsistent though :joy: JS is just different.. right now i wish we had function overloading in rust :D

**Comment by amrbashir** [comment]
Alright, added that along with a few methods on the `Builder` that should make it more Rusty hopefully.

**Review by lucasfernog — APPROVED** [review]

**Pull request merged**